server/src/main/java/org/tctalent/server/service/db/impl/DuolingoCouponServiceImpl.java

```
@@ -0,0 +1,206 @@
+      coupon.setCouponStatus(status);
+      couponRepository.save(coupon);
+    });
+  }
+
+  /**
+   * Retrieves all available (unassigned) coupons.
+   *
+   * @return a list of available Coupon entities
+   */
+  @Override
+  @Transactional(readOnly = true)
+  public List<DuolingoCoupon> getAvailableCoupons() {
+    return couponRepository.findByCandidateIsNullAndCouponStatus("AVAILABLE");
+  }
+
+  /**
+   * Finds a coupon by its unique code and maps it to a CouponResponse.
+   *
+   * @param couponCode the code of the coupon to find
+   * @return an Optional containing the mapped CouponResponse, or empty if not found
+   */
+  @Override
+  @Transactional(readOnly = true)
+  public Optional<DuolingoCouponResponse> findByCouponCode(String couponCode) {
+    return couponRepository.findByCouponCode(couponCode).map(this::mapToCouponResponse);
+  }
+
+  // Utility Methods
+
+  /**
+   * Maps a Coupon entity to a CouponResponse DTO.
+   */
+  private DuolingoCouponResponse mapToCouponResponse(DuolingoCoupon coupon) {
+    return new DuolingoCouponResponse(
+        coupon.getId(),
+        coupon.getCouponCode(),
+        coupon.getAssigneeEmail(),
+        coupon.getExpirationDate(),
+        coupon.getDateSent(),
+        coupon.getCouponStatus(),
+        coupon.getTestStatus()
+    );
+  }
+
+  /**
+   * Returns a nullable string, or null if the value is empty.
+   */
+  private String getNullableValue(String value) {
+    return value != null && !value.trim().isEmpty() ? value : null;
+  }
+
+  /**
+   * Attempts to parse a date using multiple formatters.
+   */
+  private LocalDateTime parseDate(String dateString, DateTimeFormatter... formatters) {
+    if (dateString != null && !dateString.trim().isEmpty()) {
+      for (DateTimeFormatter formatter : formatters) {
+        try {
+          return LocalDateTime.parse(dateString, formatter);
+        } catch (DateTimeParseException ignored) {}
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception has been ignored. The catch-block hides information about the stack trace. Not logging or forwarding the stack trace of an exception can mask unexpected errors and complicated debugging.

**Fix**: Use a catch block to handle the exception and write a log message declaring that the exception happened. 
[Learn more.](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Handle all exceptions by creating a log for each exception using the message thrown.
```
@@ -199,3 +199,5 @@
           return LocalDateTime.parse(dateString, formatter);
-        } catch (DateTimeParseException ignored) {}
+        } catch (DateTimeParseException ignored) {
+            log.error(&#34;DateTimeParseException is caught.&#34;, ignored);
+        }
       }

```

---

I think that this is a good suggestion from CodeGuru